Sources/Sentry/SentryFramesTracker.m

```diff
@@ -0,0 +1,134 @@
+#import "SentryFramesTracker.h"
+#import "SentryDisplayLinkWrapper.h"
+#import "SentryLog.h"
+#import "SentryOptions.h"
+#include <stdatomic.h>
+
+#if SENTRY_HAS_UIKIT
+#    import <UIKit/UIKit.h>
+
+static CFTimeInterval const SentryFrozenFrameThreshold = 0.7;
+static CFTimeInterval const SentryPreviousFrameInitalValue = -1;
+
+/**
+ * Relaxed memoring ordering is typical for incrementing counters. This operation only requires
+ * atomicity but not ordering or synchronization.
+ */
+static memory_order const SentryFramesMemoryOrder = memory_order_relaxed;
+
+@interface
+SentryFramesTracker ()
+
+@property (nonatomic, strong, readonly) SentryDisplayLinkWrapper *displayLinkWrapper;
+@property (nonatomic, assign, readonly) CFTimeInterval slowFrameThreshold;
+@property (nonatomic, assign) CFTimeInterval previousFrameTimestamp;
+
+@end
+
+@implementation SentryFramesTracker {
+
+    /**
+     * With 32 bit we can track frames with 120 fps for around 414 days (2^32 / (120* 60 * 60 *
+     * 24)).
+     */
+    atomic_uint_fast32_t _totalFrames;
+    atomic_uint_fast32_t _slowFrames;
+    atomic_uint_fast32_t _frozenFrames;
+}
+
++ (instancetype)sharedInstance
+{
+    static SentryFramesTracker *sharedInstance = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedInstance =
+            [[self alloc] initWithDisplayLinkWrapper:[[SentryDisplayLinkWrapper alloc] init]];
+    });
+    return sharedInstance;
+}
+
+/** Internal constructor for testing */
+- (instancetype)initWithDisplayLinkWrapper:(SentryDisplayLinkWrapper *)displayLinkWrapper
+{
+    if (self = [super init]) {
+        _isRunning = NO;
+        _displayLinkWrapper = displayLinkWrapper;
+
+        // If we can't get the frame rate we assume it is 60.
+        double maximumFramesPerSecond = 60.0;
+        if (@available(iOS 10.3, tvOS 10.3, macCatalyst 13.0, *)) {
+            maximumFramesPerSecond = (double)[UIScreen.mainScreen maximumFramesPerSecond];
```

Comment:
l: should we validate if this returns an invalid number? like -1
didn't find anything written here https://developer.apple.com/documentation/uikit/uiscreen/2806814-maximumframespersecond?language=objc so not sure if its needed

Reply:
Good that you raise such concerns, but this property is read-only, and I think it should always return a proper value. If it doesn't, something is clearly wrong. However you comment made me do some research and I found out, that you can change the `preferredFramesPerSecond` on the [CADisplayLink](https://developer.apple.com/documentation/quartzcore/cadisplaylink/1648421-preferredframespersecond). I think we should also take this into account. I created an issue for this: https://github.com/getsentry/sentry-cocoa/issues/1165